src/task_manager.rs

```diff
@@ -0,0 +1,335 @@
+                                    .unwrap()
+                            });
+                        }
+                    }
+                }
+            }
+            None => {
+                //Do nothing as task has completed or it has been cancelled
+            }
+        }
+    }
+
+    fn do_recipe_for_lsp_request(&mut self, lsp_request: LspRequest) {
+        match lsp_request {
+            LspRequest::TypeForPos(task_id, position) => {
+                let recipe = vec![
+                    RecipeStep::GetDefIdForPosition,
+                    RecipeStep::GetTypeForDefId,
+                    RecipeStep::RespondWithType,
+                ];
+
+                self.live_recipes.insert(task_id, recipe);
+                self.send_next_step(task_id, Box::new(position));
+            }
+
+            LspRequest::Completion(task_id, position) => {
+                let recipe = vec![
+                    RecipeStep::GetDefIdForPosition,
+                    RecipeStep::GetCompletionsForDefId,
+                    RecipeStep::RespondWithCompletions,
+                ];
+
+                self.live_recipes.insert(task_id, recipe);
+                self.send_next_step(task_id, Box::new(position));
+            }
+
+            LspRequest::Initialize(task_id) => {
+                let recipe = vec![RecipeStep::RespondWithInitialized];
+
+                self.live_recipes.insert(task_id, recipe);
+                self.send_next_step(task_id, Box::new(()));
+            }
+        }
+    }
+
+    fn message_loop(mut self) {
+        loop {
+            match self.receive_channel.recv() {
+                Ok(MsgToManager::TypeResponse(TypeResponse::DefId(task_id, def_id))) => {
+                    self.send_next_step(task_id, Box::new(def_id));
+                }
+                Ok(MsgToManager::TypeResponse(TypeResponse::Type(task_id, type_id))) => {
+                    self.send_next_step(task_id, Box::new(type_id));
+                }
+                Ok(MsgToManager::TypeResponse(TypeResponse::Completions(task_id, completions))) => {
+                    self.send_next_step(task_id, Box::new(completions));
+                }
+                Ok(MsgToManager::LspRequest(lsp_request)) => {
+                    self.do_recipe_for_lsp_request(lsp_request);
+                }
+                Ok(MsgToManager::Shutdown) => {
+                    break;
+                }
+                Err(_) => {
+                    eprintln!("Error during host receive");
+                }
+            }
+        }
+
+        self.stop();
+    }
+
+    pub fn start(self) -> std::thread::JoinHandle<()> {
+        thread::spawn(move || {
+            self.message_loop();
+        })
+    }
+
+    fn stop(self) {
+        // Join all the threads
+        let _ = self.handle_for_type_checker.unwrap().join();
+        let _ = self.handle_for_lsp_responder.unwrap().join();
+    }
+
+    pub fn start_type_checker(&mut self) {
```

Comment:
This "type checker" is basically the *compiler*, right? It seems like it will basically have the job of executing the atomic — i.e., uninterruptable — queries?